### ISSUE TITLE
Potential thread safe issue?

### ISSUE THREAD

**Issue by LoungeFlyZ** [JamesNK/Newtonsoft.Json.Schema#50]
We have a bit of code that parallelizes some validations on a large collection of objects. 

Like this  (ruleSchema is a JSchema object and there is only one of them):

```
// calc how parallel 
var opts = new ParallelOptions
{
    MaxDegreeOfParallelism = Convert.ToInt32(Math.Ceiling((Environment.ProcessorCount*0.75)*1.0))
};

Parallel.ForEach(response.Entries.Cast<SearchResultEntry>(), opts, sre =>
{
    var user = new User(sre);

    var errors = this.CheckUser(user, ruleSchema);

    if (errors.Count > 0)
    {
        // do stuff
    }
    else if (generateSuccessfulUserAuditResults)
    {
        // do other stuff
    }

});
```

CheckUser does the JSON Schema validation and looks like this:

```
private IList<ValidationError> CheckUser(User user, JSchema ruleSchema)
{
    IList<ValidationError> errors = new List<ValidationError>();

    var userToCheck = user.ToJson();
    bool valid = userToCheck.IsValid(ruleSchema, out errors);

    return errors;
}
```

In some circumstances (very infrequently) we have seen this:

```
 System.ThrowHelper.ThrowInvalidOperationException(ExceptionResource resource):16
 System.Collections.Generic.List`1+Enumerator.MoveNextRare():19
 System.Collections.Generic.List`1+Enumerator.MoveNext():72
 System.Linq.Enumerable.Single[TSource](IEnumerable`1 source, Func`2 predicate):71
 Newtonsoft.Json.Schema.Infrastructure.Validation.Validator.PopulateSchemaId(ValidationError error):36
 Newtonsoft.Json.Schema.Infrastructure.Validation.Validator.RaiseError(ValidationError error):104
 Newtonsoft.Json.Schema.Infrastructure.Validation.ValidatorContext.RaiseError(IFormattable message, ErrorType errorType, JSchema schema, Object value, IList`1 childErrors):7
 Newtonsoft.Json.Schema.Infrastructure.Validation.SchemaScope.RaiseError(IFormattable message, ErrorType errorType, JSchema schema, Object value, IList`1 childErrors):7
 Newtonsoft.Json.Schema.Infrastructure.Validation.ObjectScope.EvaluateTokenCore(JsonToken token, Object value, Int32 depth):86
 Newtonsoft.Json.Schema.Infrastructure.Validation.Validator.ValidateCurrentToken(JsonToken token, Object value, Int32 depth):116
 Newtonsoft.Json.Schema.JSchemaValidatingReader.ValidateCurrentToken():170
 Newtonsoft.Json.Schema.JSchemaValidatingReader.Read():15
 Newtonsoft.Json.Schema.SchemaExtensions.Validate(JToken source, JSchema schema, SchemaValidationEventHandler validationEventHandler):51
 Newtonsoft.Json.Schema.SchemaExtensions.IsValid(JToken source, JSchema schema, IList`1& errors):31
 Antrea.Windows.Shared.Workers.Audit.AdWorker.CheckUser(User user, JSchema ruleSchema) in C:\Users\chris\Documents\Projects\antrea-windows\antrea-windows-shared\Workers\Audit\AdWorker.cs:334
 Antrea.Windows.Shared.Workers.Audit.AdWorker+<>c__DisplayClass17_0.<Audit>b__0(SearchResultEntry sre) in C:\Users\chris\Documents\Projects\antrea-windows\antrea-windows-shared\Workers\Audit\AdWorker.cs:248
 System.Threading.Tasks.Parallel+<>c__DisplayClass27`2.<PartitionerForEachWorker>b__29():726
 System.Threading.Tasks.Task.InnerInvoke():15
 System.Threading.Tasks.Task.InnerInvokeWithArg(Task childTask):0
 System.Threading.Tasks.Task+<>c__DisplayClass0.<ExecuteSelfReplicating>b__1(Object ):134
```

Could this be threadsafe issue in PopulateSchemaId?

`Uri schemaId = Schema.KnownSchemas.Single(s => s.Schema == error.Schema).Id;`

**Comment by JamesNK** [comment]
Cool bug. The issue was threads weren't waiting until the shared cache was finished populated.

https://github.com/JamesNK/Newtonsoft.Json.Schema/commit/c0b6b6056a923a2cdcea2aa1e26a53686ec6ed66

**Comment by LoungeFlyZ** [comment]
Thanks @JamesNK ... great to hear it was relatively simple to sort out. 

Ideally it would be in a release within the next week.  The error is not super common, but would like to get it patched by the 30th with our customers. Doable?

**Comment by JamesNK** [comment]
Ok. Here is a pre-release NuGet package with the fix in it

[Newtonsoft.Json.Schema.2.0.7-build000198.zip](https://github.com/JamesNK/Newtonsoft.Json.Schema/files/488718/Newtonsoft.Json.Schema.2.0.7-build000198.zip)

**Comment by LoungeFlyZ** [comment]
Thanks James .. sorry for not ack'ing sooner on this. 

Do you know when 2.0.7 will come out publicly?

**Comment by JamesNK** [comment]
I'd like to fix one more issue before 2.0.7 - I'm waiting on a repo from a user. Once it is fixed, or If I don't get a repo in the next couple of days, then I'll push out 2.0.7.

**Comment by JamesNK** [comment]
Now available - https://www.nuget.org/packages/Newtonsoft.Json.Schema/2.0.7

**Comment by LoungeFlyZ** [comment]
Thanks!  much appreciated!

**Comment by LoungeFlyZ** [comment]
Hey @JamesNK we recently noticed that there is a corner case where this issue still pops up.  If you have multiple threads doing IsValid on a set of objects in parallel and one of those threads modifies the schema (in our case it was modifying AnyOf for one property) then this same error occurs. 

I'm not sure if this is a particular common scenario, however it would seem pretty straight forward to lock during an update?

We were able to work around this by figuring out what the changes we needed  to make ahead of time and do that prior to processing beginning, but thought i would raise it anyway.
